### PR TITLE
SALTO-2852 bug fix - nacl paths exceed length limit

### PR DESCRIPTION
**Pull request by Eldardar**
Nacl paths appears to exceed the length limit in Zendesk static resource paths (macro_attachment and brand_logo).
This PR adds trim functionality of a file name to comply with filesystem restrictions.

---

_Additional context for reviewer_

---
_Release Notes_: 
* Static files with long names would be shortened and would concatenated with the name's hash

---
_User Notifications_: 

-

### CODE REVIEW

File: packages/zendesk-adapter/src/filters/macro_attachments.ts [hunk 1]
```diff
@@ -119,7 +119,7 @@ const createAttachmentInstance = ({
   const name = elementsUtils.ducktype.toNestedTypeName(
     macro.value.title, attachment.filename
   )
-  const pathName = pathNaclCase(name)
+  const pathName = pathNaclCase(naclCase(name))
   return new InstanceElement(
     naclCase(name),
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/macro_attachments.ts:124]: nit:
```suggestion
  const naclName = naclCase(name)
  const pathName = pathNaclCase(naclName)
  return new InstanceElement(
    naclName,
```

File: packages/zendesk-adapter/src/filters/help_center_locale.ts [hunk 2]
```diff
@@ -84,7 +84,7 @@ const filterCreator: FilterCreator = ({ config, client }) => ({
         localeName,
         localeType,
         { id: locale, default: locale === defaultLocale },
-        [ZENDESK, RECORDS_PATH, LOCALE_TYPE_NAME, pathNaclCase(locale)],
+        [ZENDESK, RECORDS_PATH, LOCALE_TYPE_NAME, pathNaclCase(naclCase(locale))],
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_locale.ts:87]: nit:
```suggestion
        [ZENDESK, RECORDS_PATH, LOCALE_TYPE_NAME, pathNaclCase(localeName)],
```

**Review by netama — COMMENTED** [review]
thanks! looks good, let's add tests for each of these (just give a name with 1-2 non-ascii characters and make sure both the elem name and the path are "escaped")
we should also fix this in [brand logo](https://github.com/salto-io/salto/blob/main/packages/zendesk-adapter/src/filters/brand_logo.ts#L108)

File: packages/zendesk-adapter/src/filters/macro_attachments.ts [hunk 3]
```diff
@@ -119,7 +119,7 @@ const createAttachmentInstance = ({
   const name = elementsUtils.ducktype.toNestedTypeName(
     macro.value.title, attachment.filename
   )
-  const pathName = pathNaclCase(name)
+  const pathName = pathNaclCase(naclCase(name))
```
Comment (adamro) [inline packages/zendesk-adapter/src/filters/macro_attachments.ts:122]: Notice this is also used as the StaticFile's filepath so you are losing the extension (it's replaced with a _ in naclCase). You probably need to handle that separately.

File: packages/zendesk-adapter/src/filters/help_center_locale.ts [hunk 4]
```diff
@@ -84,7 +84,7 @@ const filterCreator: FilterCreator = ({ config, client }) => ({
         localeName,
         localeType,
         { id: locale, default: locale === defaultLocale },
-        [ZENDESK, RECORDS_PATH, LOCALE_TYPE_NAME, pathNaclCase(naclCase(locale))],
+        [ZENDESK, RECORDS_PATH, LOCALE_TYPE_NAME, elemNameToNaclCasedPath(locale)],
```
Comment (adamro) [inline packages/zendesk-adapter/src/filters/help_center_locale.ts:87]: I think the Element filePath should still be `pathNaclCase(naclCase(locale))`, only the StaticFiles should use the new fn, no?

File: packages/adapter-utils/src/nacl_case_utils.ts [hunk 5]
```diff
@@ -125,6 +126,23 @@ export const naclCase = (name?: string): string => {
   return `${cleanName}${suffixFromList(specialCharsMappingList)}`
 }
 
+export const normalizeStaticResourcePath = (name: string): string => {
+  const nameBuffer = Buffer.from(name)
+  const nameByteLength = nameBuffer.byteLength
+  if (nameByteLength <= MAX_PATH_LENGTH) {
+    return naclCase(name)
```
Comment (netama) [inline packages/adapter-utils/src/nacl_case_utils.ts:133]: why are we nacl-casing resource paths? I thought the decision was to keep it unicode? @ori-moisis did I miss something?

Comment (netama) [reply packages/adapter-utils/src/nacl_case_utils.ts:133]: also, can you please add tests? :)

File: packages/zendesk-adapter/src/filters/brand_logo.ts [hunk 6]
```diff
@@ -105,7 +105,7 @@ const getBrandLogo = async ({ client, brand }: {
   const name = elementsUtils.ducktype.toNestedTypeName(
     brand.value.name, logoValues.file_name
   )
-  const pathName = pathNaclCase(name)
+  const pathName = normalizeStaticResourcePath(name)
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brand_logo.ts:108]: we shouldn't use this for everything - there are two parts:
1. the element's path - should use `pathNaclCase(naclCase())` (the first solution)
2. static resources - should use the new truncation+hash logic

this means this filter should change a bit, the element creation should use the `pathName` but the static resource should be calculated based on the new function on the `name`

File: packages/zendesk-adapter/src/filters/help_center_locale.ts [hunk 7]
```diff
@@ -84,7 +84,7 @@ const filterCreator: FilterCreator = ({ config, client }) => ({
         localeName,
         localeType,
         { id: locale, default: locale === defaultLocale },
-        [ZENDESK, RECORDS_PATH, LOCALE_TYPE_NAME, pathNaclCase(locale)],
+        [ZENDESK, RECORDS_PATH, LOCALE_TYPE_NAME, normalizeStaticResourcePath(locale)],
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_locale.ts:87]: please use `pathNaclCase(localeName)`

File: packages/zendesk-adapter/src/filters/macro_attachments.ts [hunk 8]
```diff
@@ -119,7 +119,7 @@ const createAttachmentInstance = ({
   const name = elementsUtils.ducktype.toNestedTypeName(
     macro.value.title, attachment.filename
   )
-  const pathName = pathNaclCase(name)
+  const pathName = normalizeStaticResourcePath(name)
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/macro_attachments.ts:122]: same as above - please distinguish between the nacl paths and the static resource paths

File: packages/adapter-utils/test/nacl_case_utils.test.ts [hunk 9]
```diff
@@ -153,4 +154,37 @@ describe('naclCase utils', () => {
       })
     })
   })
+
+  describe('normalizeStaticResourcePath func', () => {
```
Comment (ori-moisis) [inline packages/adapter-utils/test/nacl_case_utils.test.ts:158]: tests should include non-ascii characters

File: packages/zendesk-adapter/src/filters/brand_logo.ts [hunk 10]
```diff
@@ -106,6 +106,7 @@ const getBrandLogo = async ({ client, brand }: {
     brand.value.name, logoValues.file_name
   )
   const pathName = pathNaclCase(name)
```
Comment (ori-moisis) [inline packages/zendesk-adapter/src/filters/brand_logo.ts:108]: is `pathName` still missing a `naclCase` or is this intentionally not doing it?
won't that potentially fail in the same way as the current failure with the static files?

Comment (netama) [inline packages/zendesk-adapter/src/filters/brand_logo.ts:108]: please use `pathNaclCase(naclCase(name))` here like we had in the first iteration

File: packages/zendesk-adapter/src/filters/macro_attachments.ts [hunk 11]
```diff
@@ -119,16 +119,18 @@ const createAttachmentInstance = ({
   const name = elementsUtils.ducktype.toNestedTypeName(
     macro.value.title, attachment.filename
   )
+  const naclName = naclCase(name)
   const pathName = pathNaclCase(name)
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/macro_attachments.ts:123]:
```suggestion
  const pathName = pathNaclCase(naclName)
```

Comment (netama) [reply packages/zendesk-adapter/src/filters/macro_attachments.ts:123]: please add tests for these (see https://github.com/salto-io/salto/pull/3434#pullrequestreview-1147299939 )

File: packages/zendesk-adapter/src/filters/brand_logo.ts [hunk 12]
```diff
@@ -120,7 +121,7 @@ const getBrandLogo = async ({ client, brand }: {
       filename: logoValues.file_name,
       contentType: logoValues.content_type,
       content: new StaticFile({
-        filepath: `${ZENDESK}/${BRAND_LOGO_TYPE.elemID.name}/${pathName}`,
+        filepath: `${ZENDESK}/${BRAND_LOGO_TYPE.elemID.name}/${resourcePathName}`,
```
Comment (ori-moisis) [inline packages/zendesk-adapter/src/filters/brand_logo.ts:124]: I think we should split the path to have `brand.value.name` as a folder name and then only `logoValues.file_name`

File: packages/adapter-utils/src/nacl_case_utils.ts [hunk 13]
```diff
@@ -125,6 +126,22 @@ export const naclCase = (name?: string): string => {
...
+export const normalizeStaticResourcePath = (name: string): string => {
+  const nameBuffer = Buffer.from(name)
+  if (nameBuffer.byteLength <= MAX_PATH_LENGTH) {
+    return name
+  }
+  const nameHash = hashUtils.toMD5(name)
+  const extIndex = nameBuffer.lastIndexOf('.')
+  // In case the file has a too long extension length or no extension at all
+  if (extIndex === -1 || nameBuffer.byteLength - extIndex > MAX_PATH_LENGTH) {
+    return nameBuffer.slice(0, MAX_PATH_LENGTH).toString().concat(`_${nameHash}`)
```
[the remark this answers is not in the document]
Comment (ori-moisis) [reply packages/adapter-utils/src/nacl_case_utils.ts:139]: I think we can use https://www.npmjs.com/package/truncate-utf8-bytes to truncate

File: packages/zendesk-adapter/src/filters/brand_logo.ts (hunk 10 above)
Comment (Eldardar) [reply packages/zendesk-adapter/src/filters/brand_logo.ts:108]: My bad, it's just my confusion. Thanks for keep tracking.

File: packages/zendesk-adapter/src/filters/brand_logo.ts (hunk 12 above)
Comment (ori-moisis) [reply packages/zendesk-adapter/src/filters/brand_logo.ts:124]: we should do this in a follow-up PR

File: packages/zendesk-adapter/src/filters/help_center_locale.ts (hunk 2 above)
Comment (ori-moisis) [inline packages/zendesk-adapter/src/filters/help_center_locale.ts:87]: nit: use `localeName` instead?
```suggestion
        [ZENDESK, RECORDS_PATH, LOCALE_TYPE_NAME, localeName],
```

**Review by ori-moisis — APPROVED** [review]

**Pull request merged**